plugin/storage/cassandra/spanstore/writer.go

```diff
@@ -127,11 +142,63 @@ func (s *SpanWriter) WriteSpan(span *model.Span) error {
 			s.tagIndexSkipped.Inc(1)
 		}
 	}
+
+	if err := s.indexTraceIDByServiceAndOperation(span.TraceID, ds); err != nil {
+		return s.logError(ds, err, "Failed to insert service and operation name into index", s.logger)
+	}
+
+	if err := s.indexSpanDuration(ds); err != nil {
+		return s.logError(ds, err, "Failed to insert duration into index", s.logger)
+	}
 	return nil
 }
 
-func (s *SpanWriter) shouldIndexTag(tag dbmodel.TagInsertion) bool {
+func (s *SpanWriter) indexSpanDuration(span *dbmodel.Span) error {
+	query := s.session.Query(durationIndex)
+	timeBucket := int((span.StartTime / (60 * 60)) / 1000000) // 1hr in microseconds TODO use config
+	var err error
+	indexByOperationName := func(operationName string) {
+		q1 := query.Bind(span.Process.ServiceName, operationName, timeBucket, span.Duration, span.StartTime, span.TraceID)
+		if err2 := s.durationIndex.Exec(q1, s.logger); err2 != nil {
+			err = err2
+		}
+	}
+	indexByOperationName("")                 // index by service name alone
+	indexByOperationName(span.OperationName) // index by service name and operation name
+	return err
+}
+
+func (s *SpanWriter) indexTraceIDByServiceAndOperation(traceID model.TraceID, span *dbmodel.Span) error {
+	startTime := span.StartTime
+	bucketNo := traceIDBucket(traceID.Low, defaultNumBuckets)
+	query1 := s.session.Query(serviceNameIndex)
+	query2 := s.session.Query(serviceOperationIndex)
+	var err error
+	q1 := query1.Bind(span.Process.ServiceName, bucketNo, startTime, span.TraceID)
+	err2 := s.serviceNameIndexMetrics.Exec(q1, s.logger)
+	if err2 != nil {
+		err = err2
+	}
+	q2 := query2.Bind(span.Process.ServiceName, span.OperationName, startTime, span.TraceID)
+	err2 = s.serviceOperationIndex.Exec(q2, s.logger)
+	if err2 != nil {
+		err = err2
+	}
+	return err
+}
 
+func traceIDHash32(traceID uint64) uint32 {
+	var buf [binary.MaxVarintLen64]byte
+	sbuf := buf[:] // below methods take a slice
+	binary.PutUvarint(sbuf, traceID)
```

Comment:
I think it's in the BigEndian. We use it to convert trace IDs to db.traceID. In fact, if you had db.traceID, it would be already a byte array.

> you'll notice the hash is only used against % numBuckets.

in the original code we used random buckets. I don't like modulo on trace ID due to how our downsampling works, a hash is better.